### ISSUE TITLE
softmax as last layer

### ISSUE THREAD

**Issue by mongoose54** [jocicmarko/ultrasound-nerve-segmentation#18]
I am trying to modify the activation layer at the end of the U-Net implementation by doing this:

>conv10 = Convolution2D(1, 1, 1, activation='linear')(conv9)
>reshapeLayer = Reshape((img_rows*img_cols, 1))(conv10)
    softmaxLayer = Activation('softmax')(reshapeLayer)
    model = Model(input=inputs, output=softmaxLayer)
 model.compile(optimizer=Adam(lr=1e-5), loss='categorical_crossentropy', metrics=[dice_coef])

Unfortunately I keep getting mismatch errors for `imgs_mask_train` . I tried 3 approaches:

1) Flattening by doing `imgs_mask_train =  imgs_mask_train.ravel()` 

2) Doing some transposing and reshaping: 
>mgs_mask_train =  np.transpose(imgs_mask_train,(0,2,3,1)) \\Move channel layer at the end
    imgs_mask_train = np.reshape(imgs_mask_train, (imgs_mask_train.shape[0], 
>imgs_mask_train.shape[1]*imgs_mask_train.shape[2], imgs_mask_train.shape[3])) 

That didn't work either.
3) Categorical conversion:
`imgs_mask_train =  to_categorical(imgs_mask_train)`

Any suggestions to fix this and do softmax? I am new to Keras and perhaps I am still confused.

**Comment by jocicmarko** [comment]
It is pointless to use softmax layer if you have only 1 output channel... You need at least 2.

**Comment by mongoose54** [comment]
Hi @jocicmarko thanks for the reply. That helped. Also by replacing dice score with categorical crossentropy as an objective function I get dice score of ~98!

For the record I modified your model architecture to get 2 classes as output:
>    conv10 = Convolution2D(2, 1, 1, activation='linear')(conv9)
    perm = Permute((2,3,1))(conv10)
    flat = Reshape((img_rows*img_cols, 2))(perm)
    soft = Activation('softmax')(flat)

The model summary at the output is the following now: https://gist.github.com/mongoose54/7b9b49959b653a913e58e31ceb8b942b

Then I did 
>    imgs_mask_train = to_categorical(imgs_mask_train, 2)
    imgs_mask_train = np.reshape(imgs_mask_train,(length,img_rows* img_cols,2))

If there is a better and more elegant way (e.g. avoiding permutation, reshaping) can you describe it? 

Thanks,
Alex

PS
That brings a discussion on the use of dice score as an objective function vs categorical crossentropy. I would be interested in analyzing the reason for this large discrepancy.

**Comment by jocicmarko** [comment]
@mongoose54 that is the exact way to do it actually, there is no more elegant way, at least not yet :).

I find it wierd that you got such a huge jump in Dice score. Did you use the same dataset (the Kaggle one I mean). I remember that during competition I also tried softmax output, which was exactly the same like yours, and I got slightly worse Dice score. Are you sure you didn't make some mistakes in the calculations? If you had 0.98 on that competition you would be 1st by a huge margin ;)

**Comment by mongoose54** [comment]
@jocicmarko lol most probably I am doing something wrong on my end since I am new to Keras. In any case I put the notebook here: http://nbviewer.jupyter.org/gist/mongoose54/2420a55ef51a8561eb6a3a2930ca4520 I am basically doing a validation split of 20%. It is also weird that the accuracy is that high. Let me know where the mistake is. I want to learn properly. Thanks again for the input.

**Comment by jocicmarko** [comment]
@mongoose54 I can't open your notebook (it is loading forever), but please note that Dice score is NOT equal to accuracy! Accuracy is not a good metric for segmentation task as there are usually huge class imbalances - like most is background (0s) and only small portion is useful stuff (1s). So say that on average every image has around 90% background and only 10% foreground - that means if you predict completely black image as a mask you would get 90% accuracy, which is very misleading!

**Comment by mongoose54** [comment]
@jocicmarko My bad! In my previous comment by "accuracy" I meant "DICE score". That is the metric I used. Sorry for the confusion. I am attaching the notebook here from gist: https://gist.github.com/mongoose54/5e8bedb589665932a4614df0bf8097f9 Hope this helps.

**Comment by jocicmarko** [comment]
@mongoose54 oh okay I finally see the error :). You are using softmax layer instead of sigmoid, and have 2 outputs (for background and foreground), so when you flatten softmax output (in the Dice loss function) you have 1s for both background and foreground. That messes up computing the Dice score for foreground. So make sure you modify the loss function accordingly, i.e. to use only foreground part of the softmax layer.

**Comment by umtbudak** [comment]
there is no np_utils for to_categorical function.
how can ı install it.
thank u
